File: queues/kafka/Python/kafkaclient.py
"""
Copyright Core DF

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Kafka helpers for Core Auto. **Produce** from step scripts; **Consume** for
[ingress](../../ingress/Python/README.md) bridges only (not steps).
"""

import json
import os
from typing import Any, Optional

from lib.result import missing_env, transport_error

_bootstrap = os.environ.get("KAFKA_BOOTSTRAP_SERVERS", "")


def _config(extra: Optional[dict] = None) -> dict:
    cfg = {"bootstrap.servers": _bootstrap}
    if os.environ.get("KAFKA_SECURITY_PROTOCOL"):
        cfg["security.protocol"] = os.environ["KAFKA_SECURITY_PROTOCOL"]
    if os.environ.get("KAFKA_SASL_MECHANISM"):
        cfg["sasl.mechanism"] = os.environ["KAFKA_SASL_MECHANISM"]
    if os.environ.get("KAFKA_SASL_USERNAME"):
        cfg["sasl.username"] = os.environ["KAFKA_SASL_USERNAME"]
    if os.environ.get("KAFKA_SASL_PASSWORD"):
        cfg["sasl.password"] = os.environ["KAFKA_SASL_PASSWORD"]
    if extra:
        cfg.update(extra)
    return cfg


def Init() -> dict:
    if not _bootstrap:
        return missing_env("KAFKA_BOOTSTRAP_SERVERS")
    return {"status_code": 200}


def Produce(topic: str, value: Any, key: Optional[str] = None) -> dict:
    if not _bootstrap:
        return missing_env("KAFKA_BOOTSTRAP_SERVERS")
    try:
        from confluent_kafka import Producer
    except ImportError:
        return {"status_code": 500, "error": "confluent-kafka package required"}

    if isinstance(value, (dict, list)):
        payload = json.dumps(value).encode("utf-8")
    elif isinstance(value, bytes):
        payload = value
    else:
        payload = str(value).encode("utf-8")

    err_holder = []

    def _acked(err, msg):
        if err:
            err_holder.append(str(err))

    try:
        producer = Producer(_config())
        producer.produce(
            topic,
            payload,
            key=key.encode("utf-8") if key else None,
            callback=_acked,
        )
        producer.flush(timeout=30)
    except Exception as exc:
        return transport_error(str(exc))

    if err_holder:
        return {"status_code": 500, "error": err_holder[0]}
    return {"status_code": 200}


def Consume(
    topic: str,
    timeout_sec: float = 30,
    max_messages: int = 1,
    group_id: Optional[str] = None,
) -> dict:
    """Poll messages from a topic. For ingress processes only — not step scripts."""
    if not _bootstrap:
        return missing_env("KAFKA_BOOTSTRAP_SERVERS")
    try:
        from confluent_kafka import Consumer
    except ImportError:
        return {"status_code": 500, "error": "confluent-kafka package required"}

    gid = group_id or os.environ.get("KAFKA_GROUP_ID") or "coreauto-step"
    cfg = _config(
        {
            "group.id": gid,
            "auto.offset.reset": os.environ.get("KAFKA_AUTO_OFFSET_RESET", "earliest"),
        }
    )
    messages = []
    try:
        consumer = Consumer(cfg)
        consumer.subscribe([topic])
        deadline = timeout_sec
        while len(messages) < max_messages and deadline > 0:
            msg = consumer.poll(min(1.0, deadline))
            deadline -= 1.0
            if msg is None:
                continue
            if msg.error():
                consumer.close()
                return {"status_code": 500, "error": str(msg.error())}
            raw = msg.value()
            try:
                body = json.loads(raw.decode("utf-8"))
            except (UnicodeDecodeError, json.JSONDecodeError):
                body = raw.decode("utf-8", errors="replace")
            messages.append(
                {
                    "topic": msg.topic(),
                    "partition": msg.partition(),
                    "offset": msg.offset(),
                    "key": msg.key().decode("utf-8") if msg.key() else None,
                    "value": body,
                }
            )
        consumer.close()
    except Exception as exc:
        return transport_error(str(exc))

    return {"status_code": 200, "messages": messages}

File: queues/kafka/Node/kafkaclient.js
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//
// Kafka helpers for Core Auto. Produce from step scripts; Consume for ingress bridges.

import { Kafka } from 'kafkajs';
import { missingEnv, transportError } from './lib/result.js';

function encode(value) {
  if (typeof value === 'object' && value !== null) return JSON.stringify(value);
  return String(value);
}

function decode(raw) {
  const text = raw.toString('utf-8');
  try { return JSON.parse(text); } catch { return text; }
}

function config() {
  const cfg = { brokers: (process.env.KAFKA_BOOTSTRAP_SERVERS || '').split(',') };
  if (process.env.KAFKA_SASL_USERNAME) {
    cfg.sasl = {
      mechanism: process.env.KAFKA_SASL_MECHANISM || 'plain',
      username: process.env.KAFKA_SASL_USERNAME,
      password: process.env.KAFKA_SASL_PASSWORD || '',
    };
    cfg.ssl = process.env.KAFKA_SECURITY_PROTOCOL?.includes('SSL') ?? false;
  }
  return cfg;
}

export function Init() {
  if (!process.env.KAFKA_BOOTSTRAP_SERVERS) return missingEnv('KAFKA_BOOTSTRAP_SERVERS');
  return { status_code: 200 };
}

export async function Produce(topic, value, key = null) {
  if (!process.env.KAFKA_BOOTSTRAP_SERVERS) return missingEnv('KAFKA_BOOTSTRAP_SERVERS');
  const kafka = new Kafka(config());
  const producer = kafka.producer();
  try {
    await producer.connect();
    await producer.send({
      topic,
      messages: [{ key: key ?? undefined, value: encode(value) }],
    });
    return { status_code: 200 };
  } catch (exc) {
    return transportError(String(exc.message || exc));
  } finally {
    await producer.disconnect().catch(() => {});
  }
}

export async function Consume(topic, timeoutSec = 30, maxMessages = 1, groupId = null) {
  if (!process.env.KAFKA_BOOTSTRAP_SERVERS) return missingEnv('KAFKA_BOOTSTRAP_SERVERS');
  const kafka = new Kafka(config());
  const consumer = kafka.consumer({
    groupId: groupId || process.env.KAFKA_GROUP_ID || 'coreauto-step',
  });
  const messages = [];
  try {
    await consumer.connect();
    await consumer.subscribe({ topic, fromBeginning: true });
    await consumer.run({
      eachMessage: async ({ message }) => {
        messages.push({
          topic,
          partition: message.partition,
          offset: message.offset,
          key: message.key ? message.key.toString('utf-8') : null,
          value: decode(message.value),
        });
        if (messages.length >= maxMessages) await consumer.stop();
      },
    });
    await new Promise((r) => setTimeout(r, timeoutSec * 1000));
    await consumer.stop();
    return { status_code: 200, messages };
  } catch (exc) {
    return transportError(String(exc.message || exc));
  } finally {
    await consumer.disconnect().catch(() => {});
  }
}
